File: problems/133/solution.go
package main
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */
func cloneGraph(node *Node) *Node {
    return clone(node, make(map[*Node]*Node))
}

func clone(node *Node, m map[*Node]*Node) *Node {
    if node == nil {
        return nil
    }
    
    if found, ok := m[node]; ok {
        return found
    }
    
    cloned := &Node{node.Val, nil}
    m[node] = cloned
    
    for _, neighbor := range node.Neighbors {
        cloned.Neighbors = append(cloned.Neighbors, clone(neighbor, m))
    }
    
    return cloned
}

File: problems/133/solution.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
    if (!node) {
        return null;
    }
    
    return helper(node, new Map());
};

function helper(node, m) {
    if (m.has(node)) {
        return m.get(node);
    }
    
    const cloned = new Node(node.val);
    m.set(node, cloned);
    
    for (const neighbor of node.neighbors) {
        cloned.neighbors.push(helper(neighbor, m));
    }
    
    return cloned;
}